erdos/src/dataflow/state.rs

```
@@ -1,43 +1,92 @@
+impl State for () {
+    type Item = ();
+
     fn commit(&mut self, _timestamp: &Timestamp) {}
 
-    fn get_last_committed_timestamp(&self) -> Timestamp {
+    fn last_committed_timestamp(&self) -> Timestamp {
         Timestamp::Bottom
     }
+
+    fn at(&mut self, _timestamp: &Timestamp) -> Option<&mut Self::Item> {
+        None
+    }
+}
+
+/// The `TimeVersionedState` provides a default implementation of the `State` for a type S.
+/// The structure automatically commits the final state for a timestamp into a HashMap and
+/// initializes new states for a timestamp `t` by invoking their default method.
+pub struct TimeVersionedState<S>
+where
+    S: 'static + Default + Send + Sync,
+{
+    state: HashMap<Timestamp, S>,
+    last_committed_timestamp: Timestamp,
+}
+
+impl<S> TimeVersionedState<S>
+where
+    S: 'static + Default + Send + Sync,
+{
+    pub fn new() -> Self {
+        Self {
+            state: HashMap::new(),
+            last_committed_timestamp: Timestamp::Bottom,
+        }
+    }
+}
+
+impl<S> State for TimeVersionedState<S>
+where
+    S: 'static + Default + Send + Sync,
+{
+    type Item = S;
+
+    fn commit(&mut self, timestamp: &Timestamp) {
+        self.last_committed_timestamp = timestamp.clone();
+    }
+
+    fn last_committed_timestamp(&self) -> Timestamp {
+        self.last_committed_timestamp.clone()
+    }
+
+    fn at(&mut self, timestamp: &Timestamp) -> Option<&mut Self::Item> {
+        Some(self.state.entry(timestamp.clone()).or_default())
+    }
 }
 
 /// Trait that must be implemented by a state structure that is used in a Sequential operator.
 /// This state structure must implement an `append` method that enables message callbacks to add
 /// intermediate state to the structure, and a `commit` method that commits the final state for a
 /// given timestamp t.
-pub trait AppendableStateT<S>: 'static + Clone + Send + Sync {
+pub trait AppendableState<S>: 'static + Clone + Send + Sync {
     fn append(&self, data: &S);
```

Does `append()` need to take a timestamp as an argument? I'm concerned about scenarios where callbacks for multiple timestamps run simultaneously. In that case, it may also need to return an error (e.g. appending to an already-committed timestamp).